### PR TITLE
64907 gcp hosting

### PR DESCRIPTION
**Pull request by jacob6838**
- Hosting NavJoy 568 translator in GCP cloud (https://console.cloud.google.com/functions/details/us-central1/navjoy-568-translator?authuser=4&project=cdot-oim-wzdx-dev)
- Created new gcp_navjoy_translator script which will download newest 568 form data, convert it to WZDx, and publish generated messages to a pub/sub topic. 
- Deployment procedure involved zipping entire project, uploading, and setting start script/function in build environment variable

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -115,12 +115,24 @@ python -m pytest 'tests/' -v
...
 
 ![alt text](translator/GCP_cloud_function/iCone%20Translator%20block%20diagram.png)
 
+#### COTrip/SalesForce
+The cotrip cloud function does not exist. It will be triggered when a new cotrip alert is placed in the CDOT RTDH cotrip-alerts-raw-oim-wzdx-integration pub/sub topic. The function will be located at [https://console.cloud.google.com/functions/details/us-central1/salesforce_data?authuser=4&folder=&organizationId=&project=cdot-oim-wzdx-prod](https://console.cloud.google.com/functions/details/us-central1/salesforce_data?authuser=4&folder=&organizationId=&project=cdot-oim-wzdx-prod)
+
+#### NavJoy 568
+The navjoy cloud function is triggered once per day by a pub/sub topic. The cloud function is located at [https://console.cloud.google.com/functions/details/us-central1/navjoy-568-translator?authuser=4&project=cdot-oim-wzdx-dev](https://console.cloud.google.com/functions/details/us-central1/navjoy-568-translator?authuser=4&project=cdot-oim-wzdx-dev)
```
Comment (payneBrandon) [inline README.md]:
```suggestion
The navjoy cloud function is triggered once per day by a pub/sub topic. The cloud function is located at [https://console.cloud.google.com/functions/details/us-central1/navjoy-568-translator?project=cdot-oim-wzdx-dev](https://console.cloud.google.com/functions/details/us-central1/navjoy-568-translator?project=cdot-oim-wzdx-dev)
```
Including the "authUser" portion here seems to cause issues. If we remove it, it forces the user to login and seems to work properly. We'll want to do the same with the above links.

File: README.md [hunk 2]
```diff
@@ -115,12 +115,24 @@ python -m pytest 'tests/' -v
...
 
 A system was created in google cloud platform to automatically translate iCone data to WZDx message. This system consists of two pubsub topics and a cloud function. A cloud scheduler automatically sends a message to a pubsub topic which triggers the cloud function. The cloud function retrieves iCone data from an ftp server (ftp://iconetraffic.com:42663) and translates to WZDx message. It validates the WZDx message with json schema and publishes the message to a pubsub topic.
 
 ![alt text](translator/GCP_cloud_function/iCone%20Translator%20block%20diagram.png)
 
+#### COTrip/SalesForce
+The cotrip cloud function does not exist. It will be triggered when a new cotrip alert is placed in the CDOT RTDH cotrip-alerts-raw-oim-wzdx-integration pub/sub topic. The function will be located at [https://console.cloud.google.com/functions/details/us-central1/salesforce_data?authuser=4&folder=&organizationId=&project=cdot-oim-wzdx-prod](https://console.cloud.google.com/functions/details/us-central1/salesforce_data?authuser=4&folder=&organizationId=&project=cdot-oim-wzdx-prod)
+
+#### NavJoy 568
```
Comment (payneBrandon) [inline README.md]: Can we add a description about the actual deployment with new `gcp_568_translator.py` in here, and what additional environment variables it requires? (`navjoy_568_endpoint` for instance is now required for this one to run).

File: gcp_568_translator.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
+import json
```
Comment (payneBrandon) [inline gcp_568_translator.py:1]: We'll want a couple unit tests for this just showing that a successful translation is published, and a failed translation results in error messages.

File: README.md (hunk 2 above)
Comment (payneBrandon) [reply README.md]: Same with the new icone deployment, can we add a section about that and what environment variables are needed?

File: gcp_icone_translator.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
+from translator import icone_translator
```
Comment (payneBrandon) [inline gcp_icone_translator.py]: Likewise in here, we'll want a few simple unit tests showing success/fail cases

File: requirements.txt [hunk 5]
```diff
@@ -9,4 +9,7 @@ python-dateutil
 time_machine
 shapely
 geopy
-pyproj
\ No newline at end of file
+pyproj
+numpy
+requests
+pytz
```
Comment (payneBrandon) [inline requirements.txt:15]: Is pytz used? I don't see any references to it other than in here.

Comment (jacob6838) [reply requirements.txt:15]: When I tried to deploy the function without this it errored out, very confusing. I think it is required for the cloud build or something

File: gcp_icone_translator.py (hunk 4 above)
Comment (jacob6838) [reply gcp_icone_translator.py]: This should have a test file, tests/gcp_icone_translator_test.py

File: README.md (hunk 2 above)
Comment (jacob6838) [reply README.md]: Just updated the README, I described the deployment process and the environment variables

File: gcp_568_translator.py (hunk 3 above)
Comment (jacob6838) [reply gcp_568_translator.py:1]: Added unit tests for successful publishing (calling publish) and failure results in no publish

File: README.md [hunk 6]
```diff
@@ -115,16 +115,111 @@ python -m pytest 'tests/' -v
...
+
+#### iCone
+The iCone cloud function is triggered once per day by a pub/sub topic. The cloud function is located at [https://console.cloud.google.com/functions/details/us-central1/auto_icone_translator_ftp?project=cdot-oim-wzdx-dev](https://console.cloud.google.com/functions/details/us-central1/auto_icone_translator_ftp?project=cdot-oim-wzdx-dev)
 
 A system was created in google cloud platform to automatically translate iCone data to WZDx message. This system consists of two pubsub topics and a cloud function. A cloud scheduler automatically sends a message to a pubsub topic which triggers the cloud function. The cloud function retrieves iCone data from an ftp server (ftp://iconetraffic.com:42663) and translates to WZDx message. It validates the WZDx message with json schema and publishes the message to a pubsub topic.
 
 ![alt text](translator/GCP_cloud_function/iCone%20Translator%20block%20diagram.png)
```
Comment (payneBrandon) [inline README.md]: This diagram seems to have been moved, but we include it below properly. Do we just want to remove the reference here?

File: README.md [hunk 7]
```diff
@@ -115,16 +115,111 @@ python -m pytest 'tests/' -v
...
+| :---------------------------- | :-----------------------------------------------------------------------------------------------------------------: | -------------------------------------------------------: |
+| unsupported_messages_topic_id |                                             unsupported_messages_cotrip                                             | pub/sub topic id to send unsupported/invalid messages to |
+| project_id                    |                                                 cdot-oim-wzdx-prod                                                  |                                           GCP project ID |
+| wzdx_topic_id                 |                                                wzdx_messages_cotrip                                                 |                          Generated WZDx pub/sub topic ID |
+| navjoy_568_endpoint           | https://proxy.assetgov.com/napi/open-api/Form568?api_key=d0c2feba6d38df6fdd284d370cbd69636f337d48&limit=1000&skip=0 |                     GCP script name at root of Work_Zone |
```
Comment (payneBrandon) [inline README.md]: Is this navjoy_568_endpoint needed by the COTrip function?

File: tests/tools/gcp_tools_test.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    assert 'invalid schema: not valid json' in str(runtimeErr.value)
+
+
+def test_get_wzdx_schema_not_exist():
+    with pytest.raises(RuntimeError) as runtimeErr:
+        gcp_tools.get_wzdx_schema('not_exist.json')
+    assert 'invalid schema: file does not exist' in str(runtimeErr.value)
+
+
+# @patch('google.cloud.pubsub_v1.PublisherClient')
```
Comment (payneBrandon) [inline tests/tools/gcp_tools_test.py]: Can we remove?

File: README.md (hunk 1 above)
Comment (jacob6838) [reply README.md]: These have been removed

File: README.md (hunk 6 above)
Comment (jacob6838) [reply README.md]: I removed this section, it is a duplicate of what is shown below

File: README.md (hunk 7 above)
Comment (jacob6838) [reply README.md]: good catch

File: tests/tools/gcp_tools_test.py (hunk 8 above)
Comment (jacob6838) [reply tests/tools/gcp_tools_test.py]: removed

**Comment by jacob6838** [comment]
@payneBrandon I believe all comments have been resolved on this task, let me know if there is anything else you want me to change!

**Review by payneBrandon — APPROVED** [review]
Looks good to me

**Pull request merged**